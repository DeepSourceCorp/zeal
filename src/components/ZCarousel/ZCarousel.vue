<template>
  <div class="z-carousel__wrapper relative h-full">
    <div class="z-carousel__container relative overflow-hidden h-full z-10">
      <slot></slot>
    </div>
    <div class="z-carousel__controls" v-show="showControls">
      <button
        class="z-carousel__control z-carousel__control--left absolute h-12 w-12 z-20 -mt-6 flex bg-ink-100 bg-opacity-75 justify-center items-center"
        @click="showPrevSlide"
      >
        <z-icon icon="arrow-left" />
      </button>
      <button
        class="z-carousel__control z-carousel__control--right absolute h-12 w-12 z-20 -mt-6 flex bg-ink-100 bg-opacity-75 justify-center items-center"
        @click="showNextSlide"
      >
        <z-icon icon="arrow-right" />
      </button>
    </div>
    <ul
      class="z-indicators flex absolute mt-4 -ml-2 w-full"
      v-show="showIndicators"
      :class="[`z-indicator--${indicatorPosition}`]"
    >
      <li
        v-for="(slide, index) in slides"
        :key="index"
        @click.stop="showSlide(index)"
        class="h-2 w-2 cursor-pointer rounded-full bg-ink-100 ml-3"
        :class="['z-indicator', { is_active: index == currentIndex }]"
      >
        <button class="z-indicator__button" />
      </li>
    </ul>
  </div>
</template>

<script>
import ZIcon from '../ZIcon'

export default {
  name: 'ZCarousel',
  components: {
    ZIcon
  },
  props: {
    activeIndex: {
      default: 0,
      type: Number
    },
    indicatorPosition: {
      default: 'left',
      type: String
    },
    showIndicators: {
      default: true,
      type: Boolean
    },
    autoSlide: {
      default: true,
      type: Boolean
    },
    autoTiming: {
      default: '2000',
      type: String
    },
    showControls: {
      default: true,
      type: Boolean
    }
  },
  data() {
    return {
      currentIndex: this.activeIndex || 0,
      autoInterval: null,
      slides: [],
      slideDirection: ''
    }
  },
  computed: {
    slidesLength() {
      return this.slides.length
    },
    reachedMaxLeft() {
      return this.currentIndex === 0
    }
  },
  methods: {
    showNextSlide() {
      this.currentIndex++
      if (this.currentIndex >= this.slidesLength) {
        this.currentIndex = 0
      }
      this.slideDirection = 'slide-right'
    },

    showPrevSlide() {
      this.currentIndex--
      if (this.currentIndex < 0) {
        this.currentIndex = this.slidesLength - 1
      }
      this.slideDirection = 'slide-left'
    },

    showSlide(slideIndex) {
      if (slideIndex > this.currentIndex) this.slideDirection = 'slide-right'
      else this.slideDirection = 'slide-left'
      this.currentIndex = slideIndex
    },

    cancelAutoSlide() {
      clearInterval(this.autoInterval)
    }
  },
  mounted() {
    this.slides = this.$children.filter((child) => child.$options.name === 'ZSlide')
    this.slides.map((slide, index) => {
      slide.index = index
    })
    if (this.autoSlide) {
      this.autoInterval = setInterval(() => {
        this.showNextSlide()
      }, parseInt(this.autoTiming))
    }
  },
  beforeDestroy() {
    this.cancelAutoSlide()
  }
}
</script>

<style lang="css">
.z-indicator.is_active {
  @apply bg-juniper;
}

.z-indicator--left {
  @apply justify-start;
}

.z-indicator--center {
  @apply justify-center;
}

.z-indicator--right {
  @apply justify-end;
}

.z-carousel__control--left {
  @apply left-0 top-50;
}

.z-carousel__control--right {
  @apply right-0 top-50;
}
</style>
