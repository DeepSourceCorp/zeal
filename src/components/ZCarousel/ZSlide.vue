<template>
  <transition :name="dir">
    <div class="z-slide" v-show="visible">
      <div class="z-slide__content">
        <slot></slot>
      </div>
    </div>
  </transition>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'

@Component({
  name: 'ZSlide'
})
export default class Slide extends Vue {
  private index: number = 0

  get visible(): boolean {
    let $parent: any = this.$parent
    this.index = $parent.slides.indexOf(this)
    return this.index === $parent.currentIndex
  }

  get dir(): string {
    let $parent: any = this.$parent
    return $parent.slideDirection
  }
}
</script>

<style lang="css">
.z-slide {
  @apply absolute w-full h-full top-0 left-0 bottom-0 right-0;
}
.z-slide__content {
  @apply w-full h-full;
}

.z-slide__content img {
  @apply object-fill w-full h-full;
}

.slide-left-enter-active {
  /* @apply animation-slide-left-enter-active; */
  animation: slide-left-in 0.5s ease-out;
}

.slide-left-leave-active {
  /* @apply animation-slide-left-leave-active; */
  animation: slide-left-out 0.5s ease-out;
}

@keyframes slide-left-in {
  from {
    transform: translateX(-100%);
  }
  to {
    transform: translateX(0);
  }
}
@keyframes slide-left-out {
  from {
    transform: translateX(0%);
  }
  to {
    transform: translateX(100%);
  }
}

.slide-right-enter-active {
  /* @apply slide-right-enter-active; */
  animation: slide-right-in 0.5s ease-in;
}

.slide-right-leave-active {
  /* @apply slide-right-leave-active; */
  animation: slide-right-out 0.5s ease-in;
}

@keyframes slide-right-out {
  from {
    transform: translateX(0%);
  }
  to {
    transform: translateX(-100%);
  }
}

@keyframes slide-right-in {
  from {
    transform: translateX(100%);
  }
  to {
    transform: translateX(0);
  }
}
</style>
