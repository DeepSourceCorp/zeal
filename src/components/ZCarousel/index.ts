export { default as ZCarousel } from './ZCarousel.vue'
export { default as ZSlide } from './ZSlide.vue'
