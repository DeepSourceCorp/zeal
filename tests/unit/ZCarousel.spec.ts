import ZCarousel from '../../src/components/ZCarousel/index'
import Vue from 'vue'

import { mount, Wrapper } from '@vue/test-utils'

describe('ZCarousel', () => {
  let mountFn: (options?: Record<string, unknown>) => Wrapper<Vue>
  let localVue: typeof Vue

  beforeEach(() => {
    mountFn = (options = {}) => {
      return mount(ZCarousel, {
        localVue,
        ...options
      })
    }
  })

  it('renders the component', () => {
    let wrapper = mount(ZCarousel)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('renders the Carousel Component with Slides', () => {
    const wrapper = mountFn({
      slots: {
        default: '<div>Some Random Content</div>'
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
