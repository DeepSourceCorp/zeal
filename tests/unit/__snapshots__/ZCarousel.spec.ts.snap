// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZCarousel renders the Carousel Component with Slides 1`] = `
<div class="z-carousel__wrapper">
  <div class="z-carousel__container">
    <div>Some Random Content</div>
  </div>
  <div class="z-carousel__controls"><button class="z-carousel__control z-carousel__control--left"><svg viewBox="0 0 24 24" fill="none" class="z-icon z-icon--arrow-left z-icon--default">
        <line x1="19" y1="12" x2="5" y2="12"></line>
        <polyline points="12 19 5 12 12 5"></polyline>
      </svg></button> <button class="z-carousel__control z-carousel__control--right"><svg viewBox="0 0 24 24" fill="none" class="z-icon z-icon--arrow-right z-icon--default">
        <line x1="5" y1="12" x2="19" y2="12"></line>
        <polyline points="12 5 19 12 12 19"></polyline>
      </svg></button></div>
  <ul class="z-indicators z-indicator--left"></ul>
</div>
`;

exports[`ZCarousel renders the component 1`] = `
<div class="z-carousel__wrapper">
  <div class="z-carousel__container"></div>
  <div class="z-carousel__controls"><button class="z-carousel__control z-carousel__control--left"><svg viewBox="0 0 24 24" fill="none" class="z-icon z-icon--arrow-left z-icon--default">
        <line x1="19" y1="12" x2="5" y2="12"></line>
        <polyline points="12 19 5 12 12 5"></polyline>
      </svg></button> <button class="z-carousel__control z-carousel__control--right"><svg viewBox="0 0 24 24" fill="none" class="z-icon z-icon--arrow-right z-icon--default">
        <line x1="5" y1="12" x2="19" y2="12"></line>
        <polyline points="12 5 19 12 12 19"></polyline>
      </svg></button></div>
  <ul class="z-indicators z-indicator--left"></ul>
</div>
`;
